Deprecate `ModelFilter/DatasetFilter`

druvdub opened the pull request:
Related to #1676 

Moved all attributes from ModelFilter class to list_models arguments and added a filter query builder to build a query. Updated list_models to take both filter as an Iterable and also update the filter list based on other arguments in the method

druvdub commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1469,23 +1526,37 @@ def list_models(
                 params = self._unpack_model_filter(filter)
             else:
-                params.update({"filter": filter})
+                params = self._build_filter_query(
```
Waiting to confirm on how to handle different scenarios for filter

Wauplin commented:
Hi @druvdub! Thanks for the PR :heart:  I've tried to answer all of your questions in https://github.com/huggingface/huggingface_hub/issues/1676#issuecomment-1945594952. Please let me know if some parts are still unclear :)

Regarding this PR, would it be possible to fix the styling? The PR change diff is very big at the moment due to some whitespaces/newlines added or removed. This makes the PR quite difficult to review. You can fix this with the following steps:
1. install the dev dependencies:
```
pip install -e ".[dev]"
```
2. run the `make style` command (which runs ruff linter and formatter + a few scripts) 
```
make style
``` 
3. (optional) once done, you should be able to run `run quality` without any error (checks styling + type annotations):
```
make quality
```

I believe the current changes were due to a "autoformat" or "format on save" feature in your IDE. Make sure to always run `make style` before committing changes :) Thanks in advance!

druvdub commented:
> Hi @druvdub! Thanks for the PR ❤️ I've tried to answer all of your questions in [#1676 (comment)](https://github.com/huggingface/huggingface_hub/issues/1676#issuecomment-1945594952). Please let me know if some parts are still unclear :)
> 
> Regarding this PR, would it be possible to fix the styling? The PR change diff is very big at the moment due to some whitespaces/newlines added or removed. This makes the PR quite difficult to review. You can fix this with the following steps:
> 
> 1. install the dev dependencies:
> 
> ```
> pip install -e ".[dev]"
> ```
> 
> 2. run the `make style` command (which runs ruff linter and formatter + a few scripts)
> 
> ```
> make style
> ```
> 
> 3. (optional) once done, you should be able to run `run quality` without any error (checks styling + type annotations):
> 
> ```
> make quality
> ```
> 
> I believe the current changes were due to a "autoformat" or "format on save" feature in your IDE. Make sure to always run `make style` before committing changes :) Thanks in advance!

Yup no worries, for some reason the makefile did not work for some reason, So I had to now manually run the ruff checks

druvdub commented on src/huggingface_hub/hf_api.py:
```diff
@@ -340,7 +340,8 @@ class CommitInfo(str):
 
     # legacy url for `str` compatibility (ex: url to uploaded file, url to uploaded folder, url to PR, etc.)
-    _url: str = field(repr=False, default=None)  # type: ignore  # defaults to `commit_url`
+    # type: ignore  # defaults to `commit_url`
+    _url: str = field(repr=False, default=None)
```
FYI: `mypy src` raises an error here. Not sure if I should address it

Wauplin replied:
```suggestion
    _url: str = field(repr=False, default=None)  # type: ignore  # defaults to `commit_url`
```
This should be on a single-line to avoid mypy complaining

druvdub commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -221,3 +236,10 @@ class ModelFilter:
+    def __post_init__(self):
+        warnings.warn(
+            "ModelFilter is being deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the `list_models` method.",
+            FutureWarning,
```
Updated the category from `DeprecatedWarning` to `FutureWarning` because 
> DeprecatedWarning: Base category for warnings about deprecated features when those warnings are intended for other Python developers (ignored by default, unless triggered by code in __main__).
i.e. you won't get a warning in interepreter if you use the class as it is, unless we update the filters for warning.

Wauplin replied:
Yes that's perfect! This is also what we use in [`./utils/_deprecate.py`](https://github.com/huggingface/huggingface_hub/blob/133bf03ebf22e33f558a3a4041e4ec4cca6e9e00/src/huggingface_hub/utils/_deprecation.py) (no need to use these utils here). Nice idea to use the `__post_init__` method to deprecate dataclasses :)

druvdub commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1654,7 +1794,22 @@ def list_datasets(
+                if isinstance(filter, str):
+                    params.update({"filter": params.get("filter", set()).add(filter)})
+                else:
+                    params.update({"filter": params.get("filter", set()).update(filter)})
```
Used set to avoid duplication in cases where a user passes both keyword argument and a filter
e.g.
```python

filter="language:en"
language="language:en"
```
can revert it to a tuple if needed

Wauplin replied:
_I think_ it's fine to keep it simple without the `set` overhead. I don't think we should optimize for users passing `list_datasets(..., filter="language:en", language="language:en")` (which is very unlikely to happen). In any case, if we the parameter is duplicated in the URL, the search query should still work so no need to worry about it.

---
(nit) to handle both raw strings and lists in such a case I would do something like this:
```py
                if isinstance(filter, str):
                    filter = [filter]
                params.update({"filter": params.get("filter", set()).update(filter)})
```
which is clearer to read IMO. Drawback is that the `filter` value is modified but that's not really an issue in this precise case.

druvdub commented:
@Wauplin I think this should be all in terms of functionality changes. Should I update READMEs and completely remove `ModelFilter/DatasetFilter`.

And finally in terms of test changes should I just update the existing tests to take arguments instead of `ModelFilter/DatasetFilter`

Wauplin commented:
Thanks for all the modifications and taking care of the special cases @druvdub! I'll have a look at it on tomorrow if that's fine with you :)

druvdub commented:
> Thanks for all the modifications and taking care of the special cases @druvdub! I'll have a look at it on tomorrow if that's fine with you :) 

Yea sounds good to me :)

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -73,6 +74,10 @@ class DatasetFilter:
     compatible with the REST API. For all parameters capitalization does not
     matter.
 
+    <Tip warning={true}>
+        The `DatasetFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the [`list_datasets`].
+    </Tip>
```
```suggestion
    <Tip warning={true}>
        The `DatasetFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to [`list_datasets`].
    </Tip>
```

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -151,6 +156,13 @@ class DatasetFilter:
+        warnings.warn(
+            "DatasetFilter is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the `list_datasets` method.",
+            category=FutureWarning,
+            stacklevel=2,
```
```suggestion
```
no need for that (all our deprecation warning use default stacklevel)

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -159,6 +171,10 @@ class ModelFilter:
     compatible with the REST API. For all parameters capitalization does not
     matter.
 
+    <Tip warning={true}>
+        The `ModelFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the [`list_models`].
+    </Tip>
```
```suggestion
    <Tip warning={true}>
        The `ModelFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to [`list_models`].
    </Tip>
```

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -221,3 +238,10 @@ class ModelFilter:
+        warnings.warn(
+            "ModelFilter is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the `list_models` method.",
+            FutureWarning,
+            stacklevel=2,
```
```suggestion
```
(same as above)

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1389,6 +1390,13 @@ def list_models(
         cardData: bool = False,
         fetch_config: bool = False,
         token: Optional[Union[bool, str]] = None,
+        library: Optional[Union[str, List[str]]] = None,
+        language: Optional[Union[str, List[str]]] = None,
+        model_name: Optional[str] = None,
+        task: Optional[Union[str, List[str]]] = None,
+        trained_dataset: Optional[Union[str, List[str]]] = None,
+        tags: Optional[Union[str, List[str]]] = None,
+        pipeline_tag: Optional[str] = None,
```
I would move those arguments up, between `author` and `search` => this way we start with all possible filters and then have the sort/direction/limit/full/etc. options.
(same below in docstring)

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1474,11 +1503,26 @@ def list_models(
                 params = self._unpack_model_filter(filter)
             else:
-                params.update({"filter": filter})
+                params = self._build_model_filter(
```
I think it's best to have all the `_build_model_filter` inside `list_models`. Once `ModelFilter` are completely removed (and therefore `_unpack_model_filter`), we will have everything in one place which will be more consistent that the current code. Same way as you handled `pipeline_tag` below

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1572,6 +1676,14 @@ def list_datasets(
         limit: Optional[int] = None,
         full: Optional[bool] = None,
         token: Optional[str] = None,
+        benchmark: Optional[Union[str, List[str]]] = None,
+        dataset_name: Optional[str] = None,
+        language_creators: Optional[Union[str, List[str]]] = None,
+        language: Optional[Union[str, List[str]]] = None,
+        multilinguality: Optional[Union[str, List[str]]] = None,
+        size_categories: Optional[Union[str, List[str]]] = None,
+        task_categories: Optional[Union[str, List[str]]] = None,
+        task_ids: Optional[Union[str, List[str]]] = None,
```
Same comment as for `list_models` (would prefer to have them between `author` and `search`).

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1654,7 +1794,22 @@ def list_datasets(
                 params = self._unpack_dataset_filter(filter)
             else:
-                params.update({"filter": filter})
+                params = self._build_dataset_filter(
```
Same comment as for `list_models` (would prefer to have everything in `list_datasets`).

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -3612,7 +3818,8 @@ def create_commit(
             create_pr=create_pr,
             num_threads=num_threads,
-            free_memory=False,  # do not remove `CommitOperationAdd.path_or_fileobj` on LFS files for "normal" users
+            # do not remove `CommitOperationAdd.path_or_fileobj` on LFS files for "normal" users
+            free_memory=False,
```
```suggestion
            free_memory=False,  # do not remove `CommitOperationAdd.path_or_fileobj` on LFS files for "normal" users
```
(something might be off with your `make style` command :thinking: )

druvdub replied:
@Wauplin haven't changed anything in the makefile, so it automatically formats these comments over the code, I'll revert those lines

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -8314,7 +8521,8 @@ def _prepare_upload_folder_additions(
     relpath_to_abspath = {
         path.relative_to(folder_path).as_posix(): path
-        for path in sorted(folder_path.glob("**/*"))  # sorted to be deterministic
+        # sorted to be deterministic
+        for path in sorted(folder_path.glob("**/*"))
```
```suggestion
        for path in sorted(folder_path.glob("**/*")) # sorted to be deterministic
```

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -8323,7 +8531,8 @@ def _prepare_upload_folder_additions(
     return [
         CommitOperationAdd(
-            path_or_fileobj=relpath_to_abspath[relpath],  # absolute path on disk
+            # absolute path on disk
+            path_or_fileobj=relpath_to_abspath[relpath],
```
```suggestion
            path_or_fileobj=relpath_to_abspath[relpath],  # absolute path on disk
```

Wauplin reviewed (commented):
Thanks for the PR @druvdub! Logic-wise it looks good to me! :fire: I left a couple of comments regarding the code + API but overall very nice one :) Once addressed, we will have to review some tests, especially if ModelFilter / DatasetFilter are used (since deprecation warning make the CI fail, on purpose). I can help on this part if needed (we have a `expect_deprecation` decorator for tests that are meant to break).

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2028). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

druvdub commented:
> Thanks for the PR @druvdub! Logic-wise it looks good to me! 🔥 I left a couple of comments regarding the code + API but overall very nice one :) Once addressed, we will have to review some tests, especially if ModelFilter / DatasetFilter are used (since deprecation warning make the CI fail, on purpose). I can help on this part if needed (we have a `expect_deprecation` decorator for tests that are meant to break).

Hey @Wauplin I have addressed all the comments so far, and hopefully I can look at the tests now.
In the meantime, is this expected? 
![image](https://github.com/huggingface/huggingface_hub/assets/59387969/e9aa4547-136c-4120-a2e7-177c53ce720f)

druvdub commented:
> Thanks for the PR @druvdub! Logic-wise it looks good to me! 🔥 I left a couple of comments regarding the code + API but overall very nice one :) Once addressed, we will have to review some tests, especially if ModelFilter / DatasetFilter are used (since deprecation warning make the CI fail, on purpose). I can help on this part if needed (we have a `expect_deprecation` decorator for tests that are meant to break).

Might need to add another decorator for classes if we plan to use `@expect_deprecation` or potentially modify the tests to handle warnings with `warning.catch_warnings`. I'll look into it

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -73,6 +74,10 @@ class DatasetFilter:
     compatible with the REST API. For all parameters capitalization does not
     matter.
 
+    <Tip warning={true}>
+        The `DatasetFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to [`list_datasets`].
+    </Tip>
+
```
(otherwise tips are not rendered properly - see https://github.com/huggingface/huggingface_hub/pull/2028#issuecomment-1948712119)

```suggestion
    <Tip warning={true}>

        The `DatasetFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to [`list_datasets`].

    </Tip>

```

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -159,6 +170,10 @@ class ModelFilter:
     compatible with the REST API. For all parameters capitalization does not
     matter.
 
+    <Tip warning={true}>
+        The `ModelFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to [`list_models`].
+    </Tip>
```
(otherwise tips are not rendered properly - see https://github.com/huggingface/huggingface_hub/pull/2028#issuecomment-1948712119)

```suggestion
    <Tip warning={true}>

        The `ModelFilter` class is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to [`list_models`].

    </Tip>
```

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -221,3 +237,9 @@ class ModelFilter:
+
+    def __post_init__(self):
+        warnings.warn(
+            "ModelFilter is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the `list_models` method.",
```
To make it work with `@expect_deprecation` in the tests

```suggestion
            "'ModelFilter' is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the `list_models` method.",
```

Wauplin commented on src/huggingface_hub/utils/endpoint_helpers.py:
```diff
@@ -151,6 +156,12 @@ class DatasetFilter:
 
+    def __post_init__(self):
+        warnings.warn(
+            "DatasetFilter is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the `list_datasets` method.",
```
To make it work with `@expect_deprecation` in the tests

```suggestion
            "'DatasetFilter' is deprecated and will be removed in huggingface_hub>=0.24. Please pass the filter parameters as keyword arguments directly to the `list_datasets` method.",
```

Wauplin reviewed (approved):
Thanks for iterating over this PR! Implementation looks good to me now :)

> In the meantime, is this expected?

Tips might not be formatted correctly if they contains special anchor (like urls). This is a known issue to solve in our parser but a workaround exists. A quick solution is to add newlines between the <Tip> tag and the text itself (see comments below) and it should fix the documentation page. Thanks for noticing!

> And similarly do we need to update the README.md files as well

AFAIK there is no mention of `ModelFilter`/`DatasetFilter` in the READMEs. No need to mention them "just" for the deprecation warning.

> Might need to add another decorator for classes if we plan to use @expect_deprecation or potentially modify the tests to handle warnings with warning.catch_warnings unless we deprecate the filter argument in list_models and list_datasets.

We can already use `@expect_deprecation` without needing to add a custom decorator. You just have to update the warning message in the code to add 'quotes'.

> what I am thinking about are the additional scenarios
where we pass the the filter keyword arguments and ModelFilter as well. Do we need them assuming ModelFilter will be removed in sometime

No need to test the cases where direct parameters + ModelFilter are passed at the same time. It's counter-intuitive + the user would have a direct deprecation warning when implementing it (+it will only be supported for a few version, as you said).

druvdub commented:
Sounds good, everything should be done now :)

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1737,25 +1737,29 @@ def test_list_datasets_no_filter(self):
         self.assertGreater(len(datasets), 100)
         self.assertIsInstance(datasets[0], DatasetInfo)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1737,25 +1737,29 @@ def test_list_datasets_no_filter(self):
         self.assertTrue("huggingface" in datasets[0].author)
         self.assertTrue("DataMeasurementsFiles" in datasets[0].id)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1737,25 +1737,29 @@ def test_list_datasets_no_filter(self):
         self.assertGreater(len(datasets), 0)
         self.assertTrue("benchmark:raft" in datasets[0].tags)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1737,25 +1737,29 @@ def test_list_datasets_no_filter(self):
         self.assertGreater(len(datasets), 0)
         self.assertTrue("language_creators:crowdsourced" in datasets[0].tags)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1766,21 +1770,25 @@ def test_filter_datasets_by_language_only(self):
         self.assertTrue("language:en" in datasets[0].tags)
         self.assertTrue("language:fr" in datasets[0].tags)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1766,21 +1770,25 @@ def test_filter_datasets_by_language_only(self):
         self.assertGreater(len(datasets), 0)
         self.assertTrue("multilinguality:multilingual" in datasets[0].tags)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1766,21 +1770,25 @@ def test_filter_datasets_by_language_only(self):
         self.assertGreater(len(datasets), 0)
         self.assertTrue("size_categories:100K<n<1M" in datasets[0].tags)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1766,21 +1770,25 @@ def test_filter_datasets_by_language_only(self):
         self.assertGreater(len(datasets), 0)
         self.assertTrue("task_categories:audio-classification" in datasets[0].tags)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("DatasetFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1858,21 +1866,25 @@ def test_list_metrics(self):
         self.assertIsInstance(metrics[0], MetricInfo)
         self.assertTrue(any(metric.description for metric in metrics))
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("ModelFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1858,21 +1866,25 @@ def test_list_metrics(self):
         self.assertGreater(len(models), 0)
         self.assertTrue("muellerzr" in models[0].modelId)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("ModelFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1858,21 +1866,25 @@ def test_list_metrics(self):
         models = list(self._api.list_models(filter=ModelFilter("facebook", model_name="bart-base")))
         self.assertTrue("facebook/bart-base" in models[0].modelId)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("ModelFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1858,21 +1866,25 @@ def test_list_metrics(self):
         models = list(self._api.list_models(filter=ModelFilter(author="muellerzr", model_name="testme")))
         self.assertEqual(len(models), 0)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("ModelFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1884,6 +1896,7 @@ def test_filter_models_with_library(self):
         )
         self.assertGreater(len(models), 0)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("ModelFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1892,11 +1905,13 @@ def test_filter_models_with_task(self):
         models = list(self._api.list_models(filter=ModelFilter(task="dummytask")))
         self.assertEqual(len(models), 0)
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("ModelFilter")
```

Wauplin commented on tests/test_hf_api.py:
```diff
@@ -1892,11 +1905,13 @@ def test_filter_models_with_task(self):
         res_en = list(self._api.list_models(filter=ModelFilter(language="en")))
         self.assertGreater(len(res_en), len(res_fr))
 
+    @expect_deprecation
```
```suggestion
    @expect_deprecation("ModelFilter")
```

Wauplin reviewed (commented):
`@expect_deprecation` requires an argument (which method/class is deprecated)

druvdub commented:
@Wauplin does the code coverage report imply I need to add tests covering the keyword arguments for filter?

Wauplin reviewed (approved):
PR is ready to be merged! Good job @druvdub :fire: :fire: :fire:
